Quadrature oscillator and complex signals

antisvin opened the pull request:
I've converted OscillatorTemplate to be a bit more versatile, so it can be used for stereo too. Arbitrary number of channels could be supported in another implementation, the only limitation is that multisignal generator doesn't have methods for per-sample processing.

Quadrature sine oscillator works and uses ~5% CPU on F4

Complex float operators are not strictly necessary here, I've used them for experiments and I guess they may end up useful. But's ok not to merge them.

antisvin committed:
Stereo oscillator template and quadrature sine

antisvin committed:
Some complex float operators

antisvin committed:
Test for quadrature sine oscillator

pingdynasty commented on LibSource/ComplexFloatArray.h:
```diff
@@ -75,6 +75,29 @@ struct ComplexFloat {
     im = magnitude*sinf(phase);
   }
 
+  float getMagnitudeSquared() {
```
you don't like `  float getMagnitudeSquared() const` on line 35?

antisvin replied:
I like it so much that I've doubled it! Ok, it probably was done by mistake. Also, I think I should mark stuff like getComplexConjugate / getDotProduct as const.

pingdynasty commented on LibSource/ComplexFloatArray.h:
```diff
@@ -75,6 +75,29 @@ struct ComplexFloat {
+  /**
+   * Copy assignment
+   */
+  ComplexFloat& operator=(const ComplexFloat& other) {
```
is this different from compiler generated default?

antisvin replied:
Right, this is probably not needed.

However, I think this struct needs a constructor that accepts a single parameter and assigns this to both members. This is because CircularBuffer does this: https://github.com/pingdynasty/OwlProgram/blob/feature/audio-processors/LibSource/CircularBuffer.h#L190 . So currently making a buffer of complex floats would raise a compilation error.

If that would be added, there's no reason not to add a constructor that assigns both members to different values. Looks like both constructors should be defined as constexpr to ensure that ComplexFloat is a literal type.

pingdynasty commented on LibSource/StereoOscillator.h:
```diff
@@ -0,0 +1,64 @@
+#ifndef __QUADRATURE_OSCILLATOR_H__
```
I think I'd rather see a ComplexOscillator

antisvin replied:
I can rename it and I've considered that name. However that would inevitably cause confusion due to common misnaming of Buchla Complex Waveform generator as "complex oscillator". OTOH, it already is the correct name for some oscillators i.e. http://www.doepfer.de/A1104.htm

So just confirm if I should rename it.

pingdynasty replied:
Yeah I meant renaming StereoOscillator => ComplexOscillator
and maybe changing that header guard...

antisvin replied:
Guard should be changed for sure. I've ended up using that name because we can have a stereo oscillator that is not outputting signal in quadrature (to be visualized in 2D). Stereo oscillator osc/template just deals with 2 channels of audio, so I've decided that a more generic name is more suitable.

But ok, I can rename it and add a comment to explain that there's no Buchla in that class ;-)

antisvin committed:
Fixes/renaming according to github discussion

antisvin committed:
Docs

antisvin commented:
Updated PR code

antisvin committed:
Complex transforms

antisvin committed:
Kill remaining inheritance from multi signal

antisvin commented:
I've added a ComplexSignalProccessor class and a family of common transform operations. I don't think FloatMatrix class should be used here as we only need to multiply a 2x2 or 3x3 matrix by a complex number (or array). In the latter case we can have faster code because 3rd output is not needed and we can reduce computation by third.

I've converted complex generators/processors to use ComplexFloatArray buffers instead of AudioBuffer. This reduces amount of boilerplate code to convert converting between audibuffer / floats / complex floats on every operation.

Also, there's now a version of quadrature sine oscillator with scalar feedback parameter. I think another one with vector value for feedback can be added too.

Test patch with some of those transforms: https://www.rebeltech.org/patch-library/patch/Complex_Transforms_test

antisvin commented:
I'll be adding tests to transform classes. And maybe it needs another class to compose multpile transforms. It should be possible to multiply matrices once per block and apply the final one instead of computing each of them sequentially (possibly with interpolation to the new matrix).

antisvin committed:
Wrong destructor call

antisvin committed:
ARM matrix support added

antisvin committed:
Matrix copy

antisvin committed:
Complex transform template

antisvin committed:
Refactored matrices

pingdynasty commented on LibSource/ComplexFloatArray.cpp:
```diff
@@ -312,3 +312,20 @@ ComplexFloatArray ComplexFloatArray::create(size_t size){
 void ComplexFloatArray::destroy(ComplexFloatArray array){
   delete[] array.data;
 }
+
```
this needs some comments to explain what it does: that it splits real and imaginary onto the first two channels

pingdynasty commented on LibSource/ComplexFloatArray.h:
```diff
@@ -99,6 +119,90 @@ struct ComplexFloat {
+
+  ComplexFloat& operator+=(float other) {
+    re += other;
+    im += other;
```
I don't think this is correct. `other` has a zero-valued imaginary part.

pingdynasty commented on LibSource/ComplexFloatArray.h:
```diff
@@ -99,6 +119,90 @@ struct ComplexFloat {
+
+  ComplexFloat& operator-=(float other) {
+    re -= other;
+    im -= other;
```
same with this

pingdynasty commented on LibSource/ComplexFloatArray.h:
```diff
@@ -2,11 +2,17 @@
 #define __ComplexFloatArray_h__
 
 #include "FloatArray.h"
+#include "AudioBuffer.h"
```
I'm not sure I like the added dependency on AudioBuffer.
How much of a convenience is it to use `cplx.copyTo(buffer);` instead of 
```
cplx.getRealValues(buffer.getSamples(0));
cplx.getComplexValues(buffer.getSamples(1))
```

This has the advantage of being very explicit.

antisvin replied:
Your second call here is supposed to be `cplx.getImaginaryValues(buffer.getSamples(1))`. 

There are a few disadvantages to this:

1. It becomes rather verbose
2. It processes data in 2 passes, so it would likely work a bit slower.

But besides not introducing an extra dependency in has another good feature - it allows swapping imaginary/real axis if necessary.

How about something like this:
```
cplx.copyTo(buffer.getSamples(0), buffer.getSamples(1));
```
I think this is better than current version and your suggestion.

Note that there already are such methods and they are used to copy data from complex array. We also need to be able to copy data into array, so maybe we should deprecate getRealValues and add copyRealTo/copyRealFrom, same for imaginary.

pingdynasty replied:
yes I like that: `copyTo(FloatArray, FloatArray)`
If you can change that and the add/subtract issues then I'll merge it.

antisvin committed:
Fixes according to github discussion

antisvin committed:
Some more docs and quadrature oscillator with complex feedback

antisvin committed:
Fixes for broken tests/warnings

antisvin commented on LibSource/Oscillator.h:
```diff
@@ -68,7 +68,7 @@ class Oscillator : public SignalGenerator {
 class OscillatorTemplate : public BaseOscillator {
 protected:
-  float mul;
+  float mul = 1;  // Prevent having NaN as default getFrequency result
```
I'm not sure if that's the best solution, but previous code was breaking oscillator test because it expected getFrequency to return 0 after constructor is called. The failure was happening here:

https://github.com/pingdynasty/OwlProgram/blob/develop/TestPatches/SineOscillatorTestPatch.hpp#L16

I suppose it may depend on compiler version used.
